### DISCUSSION TITLE
Allow Organizations/Repos to make pull requests drafts by default

### DISCUSSION THREAD

**Discussion by Xavientois** [community/community#6943] [Pull Requests]
My organization wants to try out a new flow for pull requests where most PRs should be opened as drafts rather than as ready.

It would be useful for us to be able to set up our repos to show the `Create draft pull requests` button by default rather than the `Create pull request` when opening a new PR.

![Screen Shot 2021-10-28 at 4 39 30 PM](https://user-images.githubusercontent.com/34867186/139332575-425e0104-5924-4004-831c-44e43aa8ce82.png)

Based on the existence of [this extension](https://github.com/jmatsu/default-draft-pr) and [this issue](https://github.com/refined-github/refined-github/issues/2238) shows that there as least some level of demand for configurability here.

**Comment by brianjmurrell** [comment]
And this [forum posting](https://github.community/t/set-repo-level-default-for-create-draft-pr/123218).

The particular use-case for this is organizations that have CODEOWNERS.  Adding code owners when the PR is initially opened is premature.  It could go through lots of iterations before it's ready for review, so code owners don't actually know when this is.

However since draft PRs don't add CODEOWNER reviewers, this confusion is avoided and when the PR is finally ready for review it can be converted to non-draft PRs at which time CODEOWNERS would be added.

**Reply by bigbob78C** [reply]
Thank you very much

**Comment by konstantintieber** [comment]
+1 for this.

We are relying on draft pull requests to reduce the GitHub Action minutes we spend in a large monorepo in order to save computing time and money.

Right now we include a message reminding developers to create PRs as "draft" in the pull request template, but automating this would be much better.

I wrote about this here: https://engineering.leanix.net/blog/halve-your-github-actions-bill/

**Comment by jessicajaniuk** [comment]
This would be a huge help for us, as well. Maintaining a monorepo using PullApprove pings for reviews immediately after a PR is opened. Once that PR is converted to draft, the requests are still present and maintained. It creates a bunch of noise. Defaulting to creating a draft PR would be wonderful because so many people forget to create as draft until after they clicked to create PR.

**Reply by dylhunn** [reply]
+1. We often have PRs that are opened and abandoned in a non-obviously incomplete state. It would be great if these PRs didn't consume review bandwidth.

**Comment by xavidop** [comment]
@Xavientois and I have open-sourced A Github Action that sets a PR to Draft when it is opened: https://github.com/marketplace/actions/mark-pr-as-draft

**Reply by kyleknighted** [reply]
@xavidop this appears to change Open PRs to Draft state. This, unfortunately, doesn't really solve the problem because once the PR is in Open state all the automation already happens.

**Comment by strativd** [comment]
+1

I know this is a silly use-case but PR-drafts-by-default would help alleviate a common problem with the Github/Slack integration where Pull Request previews show too much detail in Slack. Draft PRs do not show the same amount of detail, so our team will default to drafts and then on the PR right away.

- https://github.com/integrations/slack/issues/564#issuecomment-1197287546

**Comment by kevkev300** [comment]
Yes pls 😍 
we would also need that. Having quite a lot actions running on pushes to an open PR, the bill grows and grows and often on creation of the PR, it isn't in a "finished" state.

Is this already worked on or planned or something?

**Comment by jtoeri** [comment]
We are currently migrating some projects over to GitHub and are also looking for ways to configure draft to be the default for all the reasons mentioned above.

**Comment by divineunited** [comment]
Commenting as we are looking to do this for our org as well.

**Comment by LyleDavis** [comment]
Our case for this is for deployments. 

To deploy, developers raise a PR into `release` and when it's merged it deploys - we automatically run a github action that pings the relevant release approvers for that particular change when the PR is raised but skip draft PRs because they're still being worked on. Lots of devs end up raising normal PRs and then switching them to draft because it's not the default, so it would be really helpful to configure it 🙏

**Comment by seanmaisch** [comment]
I would love this for our org as well. Tired of the extra click to set things to draft for every new PR.

**Reply by lobsterkatie** [reply]
Check out Refined GH: https://github.com/refined-github/refined-github

It pulls the draft option out into its own button, so it's a one-click operation just like opening a non-draft PR: https://user-images.githubusercontent.com/34235681/152473201-868ad7c1-e06f-4826-b808-d90bca7f08b3.png

(It also does a million other very nice/useful/time-saving things. Scroll down in the repo's readme to see the list. And no, I'm not affiliated with the author in any way, just a HUGE fan of his extension.)

**Reply by brianjmurrell** [reply]
Fair enough.  But this only applies a personal preference.  I am looking to set policy for all members of an organisation or all pull requests for a repo without having to have users "cooperate" and install a browser plugin and configure it accordingly.

**Comment by jbrodie** [comment]
I can't agree with this more!  It would be great to be able to set this to the default for all repos in an Enterprise Org to be honest.  We don't need all our actions working and running when someone may just be committing code to allow someone else to be able to see the changes.  The fact that we have actions running when the commit is there is costly when we can't seem to get people to make sure they are DRAFT status to prevent CI to run.

**Comment by richard-fine** [comment]
+1 on this request - I would actually like an option to completely disable the ability to create non-draft PRs on a repo. For the few situations where someone really does want to create a non-draft PR immediately (e.g. bot-generated PRs) then it should not be difficult for the user in that situation to add the extra step of converting the PR from draft to ready.

**Comment by eL-HaXo** [comment]
+1 on this request. We run extensive automation tests on branches in a large monorepo prior to merging to main. We have our ci configured to only run against non-draft PRs and have to tell developers to remember to open PRs as drafts until they are ready to run the full automation suite against them.

**Comment by DziubaR** [comment]
+1

**Comment by rickleijten-calvi** [comment]
+1

**Comment by rafaelveggi** [comment]
+1

**Comment by AFMiziara** [comment]
+1

**Comment by jthiesen** [comment]
+1

**Comment by albertocerrone** [comment]
+1

**Comment by othman-albakri** [comment]
+1

**Comment by bazinga012** [comment]
Use [github cli](https://cli.github.com/manual/gh_pr_create) for this, you can pass `--draft` flag to mark PR draft while creation.

**Reply by albertocerrone** [reply]
How do you enforce this on a team level?

**Reply by brianjmurrell** [reply]
@bazinga012 Thanks.  But this discussion is **not** about how to open PRs as drafts. We all know how to do this already. This discussion is about an organization (or even repo) switch/option to make PRs open as drafts by default for everyone.

**Comment by meiram-tr** [comment]
In addition to the above suggestion:

## Implementation Suggestion
- Introduce a new repository-level setting in the GitHub repository settings page, allowing repository owners to enable or disable the automatic creation of pull requests as drafts.
- Update the pull request creation process to respect the repository-level setting and create new pull requests as drafts when the setting is enabled.

**Comment by acx-kesand** [comment]
This would be very beneficial to our use-case as well - we would like pre-merge checks to take place before CODEOWNER notifications are sent out.

**Comment by alebrumattinu** [comment]
+1

**Comment by roeezolantz** [comment]
+1, would really really help us to manage our CRs better, cuz we have automations that runs on ready-for-review PRs and nudge us to review them :)

**Comment by awhisler** [comment]
I feel like someone already mentioned this but curious what I'm missing and why it was overlooked.

Couldn't you just use the github action from: https://github.com/orgs/community/discussions/6943#discussioncomment-2592012 and have this be the very first thing in the pipeline. So once its opened, it basically immediately goes to Draft?

**Reply by brianjmurrell** [reply]
Did you see the reply on the suggestion you have referred to?

**Reply by awhisler** [reply]
I did but I didnt understand it until literally just now as I tried my own suggestion and see the problem haha! Shoot!

**Reply by awhisler** [reply]
I'm wondering if there's a way to do it through depending on another yml file..

**Reply by awhisler** [reply]
@brianjmurrell If you're curious, I think I found a solution!

So I use the [draft script](https://github.com/orgs/community/discussions/6943#discussioncomment-2592012) on `opened` types., then my automation tests dont run until you click 'ready for review' and it's in that type
<img width="320" alt="image" src="https://github.com/community/community/assets/2396250/60e83300-1d12-4732-a17b-6c92d860ef89">
its not quite the same as the original request but its a work around to try to not have things run automatically

**Comment by przemyslaw-survicate** [comment]
+1, great idea

**Comment by JVarga21** [comment]
+1

**Comment by jedgodsey** [comment]
+1 this should be simple and it provides obvious utility.  I think Github should make it happen.

**Comment by janwijbrand** [comment]
+1

**Comment by Huweicai** [comment]
Yes, it's really needed.

**Comment by Nocture** [comment]
+1

**Comment by l-kldzj** [comment]
+1

**Comment by marcosmamg** [comment]
+1

**Comment by TheSalarKhan** [comment]
+1

**Comment by Dale-777** [comment]
+1

**Comment by KaterinaFixter** [comment]
+1

**Comment by inginging** [comment]
+1

**Comment by ujazdowskip** [comment]
+1

**Comment by onderceylan** [comment]
Very much needed!

**Comment by ygrenzinger** [comment]
So much and should be easy, non ? :D

**Comment by opajonk** [comment]
Yup, needed!

**Comment by Drowze** [comment]
+1

**Comment by FaBrand** [comment]
Yes please :)

**Comment by guys-neosec** [comment]
+1

**Comment by eyalerez** [comment]
+1

**Comment by igor-na-walnut** [comment]
+1

**Comment by luke-brooks** [comment]
+1

**Comment by alex-huot-098** [comment]
+1

**Comment by dimmyjr-nu** [comment]
+1

**Comment by DanBotero** [comment]
+1

**Comment by tamoyal** [comment]
+1

**Comment by cbeauchesne** [comment]
+1

**Comment by Tello-Wharton** [comment]
+1

**Comment by mikeuv** [comment]
+1

**Comment by EddyBorja** [comment]
+1

**Comment by mifrej** [comment]
+1

**Comment by zameschua** [comment]
+1

**Comment by 0xRobin** [comment]
+1

**Comment by caileyfitzgerald** [comment]
+1

**Comment by kswilliames** [comment]
+1

**Comment by tariksbl** [comment]
Lots of folks waiting three years for a pretty simple feature..

**Comment by thclark** [comment]
This would be a big deal to help automation of GitHub Actions based on the event triggers related to PR readiness. Also, as a senior developer I see a couple of dozen PRs that are "ready for review" each day, none of which are ready for review. Looking at the list of PRs on a repo and knowing that if it's "ready for review" it actually *is* ready for review would be a major bonus!

**Comment by thangnv2212** [comment]
+1

**Comment by Mark-J-Lawrence** [comment]
+1

**Comment by johannes-do** [comment]
+111

**Comment by cbeauchesne** [comment]
A small fact, if ever it could help to understand that request : 

We're working on a repo that receives many contributions. The CI is quite big and can't be executed locally, so users needs to create PR to validate that the CI is ok before asking reviews. 

So, as reviewers, we're using the draft state to know that a PR is not yet ready to be reviewed. I guess it's the exact purpose of "draft".

But as users may not read carefully the PR template, or even can forget to create them as draft. Every PR in that situation asks us to look at the PR, makes useless PR review, or need to contact the users to validate that he's really requiring a review while the CI being in a bad shape. So we loose some time. 

I guesstimate the time loss per PR to be around 5 and 10 minutes, depending on the PR size. And on repos I manage, it happens two or three time each day.

So this option will save me between 10 mn and 30 mn per day. I will really appreciate that saving 😉 .

**Comment by talf** [comment]
+1

**Comment by na-jakobs** [comment]
This would be very helpful to have as an option for repositories.

**Comment by AKhozya** [comment]
+1 one on this, it annoys reviewers and doesn't allow to understand time "PR ready for review > merged"

**Comment by 111483** [comment]
+1

**Comment by LiangquanLi930** [comment]
+ 1

**Comment by taralbass** [comment]
+1 for reasons already mentioned by others. Genuinely puzzled by the longstanding, obviously popular, seemingly easy request that does not have any response from github that I saw (sorry if I missed it). Is there a reason this can't be done?

**Comment by tobiasludwig-tomtom** [comment]
+1

**Comment by ShaunMWallace** [comment]
+1 if for no other reason than to delete my "DoNotMerge" label :)

**Comment by Anthony-Beuchey-Adeo** [comment]
We also would need this feature to disallow some automatic actions while still in draft and trigger them only when ready to review. With this feature, we wouldn't have to manually set it to draft everytime, and the concerned automatic actions would not trigger at the init of the PR.

**Reply by bigbob78C** [reply]
This is great

**Comment by jimpriest** [comment]
+1 for this request...     2025 and no response from Github   :\

**Reply by bigbob78C** [reply]
We do apologize for the overlooked information.

**Comment by dan-kez** [comment]
Also tossing my hat in the ring for this feature. I help lead a devinfra department for a larger engineering organization. We would love this functionality as there is a high rate of PRs being opened in our organization that should be draft. What this results in for us is CODEOWNERS being assigned by github, webhooks for alerts going out, and churn on reviews.

We would strongly prefer to default PRs to open as a draft which can mitigate this auto assignment and noise before the PR is ready for review.

**Comment by rebelliard** [comment]
+1

**Comment by NaeemAbdullahAkram** [comment]
GitHub doesn’t have a built-in option to make all PRs drafts by default. However, you can try:

GitHub Actions: Use an action to automatically convert new PRs to drafts. Example workflow:

`name: Convert PR to Draft

on:
  pull_request:
    types: [opened]

jobs:
  convert-to-draft:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v2
      - name: Convert to Draft
        uses: peter-evans/draft-pull-request@v2
        with:
          token: ${{ secrets.GITHUB_TOKEN }}`



Manual Process: Instruct team members to create draft PRs manually by clicking “Create draft pull request” when opening a new PR.

Browser Extension: Use extensions like GitHub Pull Request Draft for easy drafting.

Feature Request: Request this feature from GitHub if it’s important.

**Comment by mb-makhlouf** [comment]
I really appreciate having this feature, it allows having our slack notifications much more productive. Our usecase is to only notifiy reviews when the developer makes his PR ready yo review

**Comment by mishaelajay** [comment]
+1

**Comment by ckittel** [comment]
+1. We're a team at Microsoft, and we'd like this feature as well.

**Comment by Super8film87** [comment]
+1

**Comment by alexbondar92** [comment]
+1

**Comment by kaushall** [comment]
+1

**Comment by joe-bader-resilience** [comment]
+1

**Comment by PierceLovesee** [comment]
+1

**Comment by erick-siordia** [comment]
+1

**Comment by lopotras** [comment]
+1

**Comment by qmg-jnguele** [comment]
+1

**Comment by tiagoasousa** [comment]
+1

**Comment by phammer** [comment]
+1

**Comment by atoulme** [comment]
+1

**Comment by ghost** [comment]
+1

**Comment by celsocoutinho-tangany** [comment]
+1

**Comment by umaralam-lt** [comment]
+1

**Comment by basabin54** [comment]
+1

**Comment by dventura3** [comment]
+1

**Comment by marcdix** [comment]
+1 (running Claude Code code review when switching from Draft => Ready)

**Comment by cschiewek** [comment]
+1

**Comment by SBTestuserCH** [comment]
+1

**Comment by wsgomes** [comment]
+1

**Comment by khaosdoctor** [comment]
+1

**Comment by mastasiano** [comment]
+1

**Comment by dan2kx** [comment]
+1

**Comment by philippe-tonic** [comment]
+1

**Comment by dandy-brandon** [comment]
+1

**Comment by alexbfree** [comment]
+1

**Comment by Jessedev1** [comment]
+1

**Reply by fredericouellet-eaton** [reply]
+1
Also if some file is in .github dir

**Comment by tcelestino** [comment]
Come on, GitHub team! This is a great feature. Prioritize this instead of adding AI everywhere.

**Comment by ecclesm** [comment]
+1 for this feature

**Comment by MikhailKandel** [comment]
+1, massive waste of time apologizing to people that drafts aren't ready for review

**Comment by dzistl** [comment]
+1

**Comment by Luisgustavom1** [comment]
+1

**Comment by jimpriest** [comment]
5 years later... maybe when they stop adding AI features they can get back to the basics and add stuff like this :\

**Comment by jakub-g** [comment]
I'm a maintainer of a massive enterprise repo (~1000 collaborators of various seniority levels). 

I want to enforce opening PRs as drafts to prevent pinging sometimes _dozens of teams_ on unfinished PRs.

But most importantly, **I want to create a bit more friction** before PRs are marked ready for review.

I want people to sit down and think, and be more accountable about their **AI-generated vibe coded PRs**:
- review the PR description
- do a self-review of the diff
- ask AI Agent for first round of review
- etc.

**Reply by AFMiziara** [reply]
Precisely. I wanted to do exactly the same at my company's monorepo.

**Comment by vkucera** [comment]
+1

**Comment by professor** [comment]
+1 very helpful with claude code opening PRs. They need to be in draft

**Comment by inloox-konrad** [comment]
We just migrated from Azure DevOps to GHE and this is not a feature? In a larger org, having PRs in a draft state by default is incredibly useful, I can only second this.

**Comment by gvsandeep2647** [comment]
+1 I think this request is now much more relevant in the age of AI. The PR volume has increased, in some case people see their code directly on GitHub and then iterate by adding comments to the PR and asking the AI to resolve it. In such scenarios the first time you have raised a PR means nothing. It's not ready for reviews / checks. It's for your preview for you to further iterate. Also, with more and more organizations adopting automated reviews, we would want them to trigger only for `ready-to-review` PRs.

**Comment by dough654** [comment]
I work at a large org and this has been a pain point for us as well. Just throwing my support in for this too.

**Comment by kenl-jfit** [comment]
Yes, +1 _specifically_ for holding off on CI until copilot reviews and resolutions are complete.

**Comment by samklevin** [comment]
+1

**Comment by atoulme** [comment]
If you are a maintainer please consider upvoting https://github.com/community/maintainers/discussions/757#discussioncomment-16716093

**Comment by tlaurion** [comment]
+1

@atoulme 
> https://github.com/community/maintainers/discussions/757#discussioncomment-16716093

I got 404

**Reply by atoulme** [reply]
you need to sign up as a github maintainer here: https://maintainers.github.com/

**Comment by Magvin** [comment]
+1

**Comment by GaryTyre** [comment]
+1

**Comment by jabenedicic** [comment]
With the increase in PRs created from all sources, it seems that now more than ever is the time to be able to have an option or policy that can be set to default PRs to draft status. It helps with so many controls that have already been mentioned in this thread.

**Comment by tejasbagal1** [comment]
+1

**Reply by Abishek-kk** [reply]
what +1

**Reply by tejasbagal1** [reply]
@Abishek-kk My organization manages around 500 repositories, with each repository using IBM Cloud DevSecOps Tekton CI pipelines.

A common development practice is for developers to create pull requests early in the development lifecycle so that progress is visible and collaboration can begin sooner. However, this results in the CI pipeline being triggered for every commit pushed to the pull request, even when the work is still incomplete. These pipeline executions often provide little value at that stage and lead to unnecessary consumption of compute resources.

If GitHub provided an option to create new pull requests as Draft by default, we could configure our pipelines to skip builds for draft pull requests. This would significantly reduce unnecessary pipeline runs, optimize resource utilization, and lower infrastructure costs while still allowing teams to track development progress.

**Comment by alexisymbiotic** [comment]
+1

**Comment by dlancy-kroger** [comment]
+1 

This request is over 5 years old and seems like such a simple thing to implement. Being able to choose what state our PR starts in would be so helpful and huge for our teams.

**Reply by jimpriest** [reply]
Maybe it should be re-worded to include the word "AI" then it might get some attention :\

**Reply by dlancy-kroger** [reply]
I could write it as we want it so AI recognizes commits in draft are still work in progress and should not be reviewed yet, and then have AI review once it changes state. 🤷

**Comment by TrimVis** [comment]
+1

**Comment by tadnir** [comment]
+1 

even just a "default open PRs as draft" in org settings would be good
